pkg/ebpf/c/tracer-ebpf.c

```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
+// All structs referenced here are kernel independent as they simply map protocol headers (Ethernet, IP and UDP).
+SEC("socket/dns_filter")
+int socket__dns_filter(struct __sk_buff* skb) {
+    skb_info_t skb_info;
+
+    if (!read_conn_tuple_skb(skb, &skb_info)) {
+        return 0;
+    }
+
+    if (skb_info.tup.sport != 53 && (!dns_stats_enabled() || skb_info.tup.dport != 53)) {
         return 0;
+    }
 
     return -1;
 }
 
+static __always_inline void http_end_response(http_transaction_t *http) {
+    if (http->state != HTTP_RESPONDING) {
+        return;
+    }
+
+    u32 cpu = bpf_get_smp_processor_id();
+    // This entries are pre-allocated in userspace
+    http_batch_t *batch = bpf_map_lookup_elem(&http_enqueued, &cpu);
+    if (batch == NULL) {
+        return;
+    }
+
+    if (batch->pos >= HTTP_BATCH_SIZE) {
+        // We keep incrementing this so we can track how many transactions we're dropping
+        batch->pos++;
+        return;
+    }
+
+    int batch_slot = (batch->idx%HTTP_BATCH_PAGES)*HTTP_BATCH_SIZE + batch->pos;
+
+    // I haven't found a way to avoid this unrolled loop on Kernel 4.4 (newer versions work fine)
+    // If you try to directly write the desired batch slot by doing
+    //
+    //  __builtin_memcpy(&batch->txs[batch_slot], http, sizeof(http_transaction_t));
+    //
+    // You get an error like the following:
+    //
+    // R0=inv R1=map_value(ks=4,vs=4816) R2=imm5 R3=imm0 R4=imm0 R6=map_value(ks=48,vs=96) R7=imm1 R8=imm0 R9=inv R10=fp
+    ///809: (79) r2 = *(u64 *)(r6 +88)
+    // 810: (7b) *(u64 *)(r0 +88) = r2
+    // R0 invalid mem access 'inv'
+    //
+    // This is because the value range of the R0 register (holding the memory address of the batch) can't be
+    // figured out by the verifier and thus the memory access can't be considered safe during verification time.
+    // I tried different "tricks" to hint the verifier of this range such as:
+    //
+    // * Ensuring 0 <= batch_slot < HTTP_BATCH_SIZE*HTTP_BATCH_PAGES
+    // * Ensuring that &batch <= &batch_slot->txs[batch_slot] <= &batch+1
+    // * Setting HTTP_BATCH_SIZE*HTTP_BATCH_PAGES to a power of 2 and doing &batch->txs[batch_slot&(HTTP_BATCH_SIZE*HTTP_BATCH_PAGES-1)]
+    //
+    // Among other things, but nothing really worked on Kernel 4.4
+    // It seems that indeed support for this type of access by the verifier was added later on:
+    // https://patchwork.ozlabs.org/project/netdev/patch/1475074472-23538-1-git-send-email-jbacik@fb.com/
+    //
+    // What is unfortunate about this is not only that enqueing a HTTP transaction is O(HTTP_BATCH_SIZE*HTTP_BATCH_PAGES),
+    // but also that we can't really increase the batch/page size at the moment because that blows up the eBPF *program* size
+#pragma unroll
+    for (int i = 0; i < HTTP_BATCH_PAGES*HTTP_BATCH_SIZE; i++) {
+        if (i == batch_slot) {
+            __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
```

Comment:
I wonder if we could use one of the atomic intrinsics to avoid writing to txs[i] while userspace reads it. 

So userspace could write FREE to txs[i] and the kprobe could do  

```
if ! ( compare_and_swap(txs[i], FREED, WRITING) ) {
    // increment counter for dropped HTTP event
} else {
   __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
}
```

Reply:
nevermind, this relies on being able to write to part of the batch from userspace which isn't possible